### PR TITLE
ToggleGroupControl: improve animation

### PR DESCRIPTION
**Pull request by DaniGuardiola**
(https://github.com/WordPress/gutenberg/pull/64926 needs to merge first)

## What?

Replaced framer motion based animation with a faster CSS animation similar to the work already done in Tabs (see https://github.com/WordPress/gutenberg/pull/64926 for latest update there).

## Why?

Faster and more lightweight, also see https://github.com/WordPress/gutenberg/issues/60975

## How?

Implement similar animation to Tabs, with the main difference being that instead of being transform-based, it is left + width based since that retains appropriate border radius dimensions.

## Testing Instructions

Try the component in the Storybook or in Gutenberg (e.g. in the block inspector with a paragraph selected -> "Typography - SIZE").

### Testing Instructions for Keyboard

Try the component with a keyboard.

## Screenshots or screencast 

Looks exactly the same as before, possibly feels a bit more snappy though it's hard to tell.

https://github.com/user-attachments/assets/a716ef6e-81c1-466d-b385-411b11820894

### CODE REVIEW

**Comment by mirka** [comment]
I'm done for the day so I haven't reviewed yet, but noting that this is also a bug fix for animation issues like this ✨ 

https://github.com/user-attachments/assets/0213220f-a07b-422e-8593-59d9cefd4035

(Seen in the Query block when "Inner blocks use content width" is enabled.)

**Comment by DaniGuardiola** [comment]
@mirka nice! I didn't know about that bug, glad to learn that it is fixed by this :)

File: packages/components/src/toggle-group-control/toggle-group-control/component.tsx
Comment (ciampo) [inline packages/components/src/toggle-group-control/toggle-group-control/component.tsx]: Nit: Wondering if relying on `querySelector` and `parentElement`is too brittle, as it could break if the underlying DOM tree changes.

An alternative could be to expose `setActiveElement` via internal context to the option sub-components, and have them call `setActiveElement` with the relevant HTML element.

File: packages/components/src/toggle-group-control/toggle-group-control/component.tsx
Comment (ciampo) [inline packages/components/src/toggle-group-control/toggle-group-control/component.tsx]: Maybe better to add a code comment explaining that we're using the `pseudoElement` property to make sure we're targeting the right transitionend event?

File: packages/components/src/toggle-group-control/toggle-group-control/styles.ts
Comment (ciampo) [inline packages/components/src/toggle-group-control/toggle-group-control/styles.ts]: Could we switch to using `transform` instead, like we recently did with `Tabs`?

File: packages/components/src/toggle-group-control/toggle-group-control/styles.ts
Comment (ciampo) [inline packages/components/src/toggle-group-control/toggle-group-control/styles.ts:42]: Let's take this opportunity to use `COLORS.theme.gray` instead

**Review by ciampo — COMMENTED** [review]
When using the arrow keys to quickly change the selected tab, the animation is sometimes "skipped" and the backdrop just jumps to the newly selected option:

https://github.com/user-attachments/assets/0e61a83a-fb65-4aa0-b705-377fea3ed077

Also, the backdrop is not showing at all when the `isDeselectable` prop is `true`: 

https://github.com/user-attachments/assets/626a7198-9402-4910-9fdc-158025be1026

**Review by tyxla — COMMENTED** [review]
Was just testing this and noticed that deselectable toggles are broken:


https://github.com/user-attachments/assets/1b9230d8-a397-4278-ab64-25795aed3384

Edit: just saw @ciampo reported this earlier - sorry for the noise!

File: packages/components/src/toggle-group-control/toggle-group-control/styles.ts
Comment (DaniGuardiola) [reply packages/components/src/toggle-group-control/toggle-group-control/styles.ts]: That was the original implementation, but unfortunately it distorts border radius values. You can play with my animation sandbox here to see what I mean: https://stackblitz.com/edit/solidjs-templates-grt7kw?file=src%2FApp.tsx

File: packages/components/src/toggle-group-control/toggle-group-control/component.tsx
Comment (DaniGuardiola) [reply packages/components/src/toggle-group-control/toggle-group-control/component.tsx]: Done in fe2b51e9f17be939d789c4f7dc14ec6771876702, it fixes the `isDeselectable` issue as well ✨

File: packages/components/src/toggle-group-control/toggle-group-control/component.tsx
Comment (DaniGuardiola) [reply packages/components/src/toggle-group-control/toggle-group-control/component.tsx]: Hmm, I kinda feel like it's somewhat self-explanatory by the context, especially if you take a quick look at the styles. I'm not opposed to adding a comment but it does seem like it'd be a bit redundant to me.

File: packages/components/src/toggle-group-control/toggle-group-control/styles.ts
Comment (DaniGuardiola) [reply packages/components/src/toggle-group-control/toggle-group-control/styles.ts:42]: Done in fe2b51e9f17be939d789c4f7dc14ec6771876702

**Comment by DaniGuardiola** [comment]
Thanks @ciampo and @tyxla. Missing indicator bug fixed in fe2b51e9f17be939d789c4f7dc14ec6771876702, along with a lot of polish and fixes.

Regarding the skipped animations, I'm trying to debug and it's a bit of a head-scratcher. Appreciate any suggestions in the meantime! Not sure what I'm missing...

**Comment by DaniGuardiola** [comment]
Well, that was an annoying rabbit hole. Without getting too deep into the details (which I don't know in much depth anyway), the issue was related to the fact that when an element changes, the time between that instant and the moment when the indicator position state is updated after the resize observer fires its first event observing the new element is... non-zero.

When updating the value at a very specific moment (somewhere around the end of the transition), it caused an obnoxious race condition that resulted in the reported skipping behavior. The fix is to track the element in sync with the rect observations, for which I have added a property to the stateful object returned by the utilities, and it seems to solve the whole thing.

That was a mouthful! Phew.

I backported the fix to Tabs, though it's interesting that we didn't see the issue there. Probably because it's just more obvious in this component.

PR ready to go?

**Comment by DaniGuardiola** [comment]
Hmm one last issue to fix with the context approach @ciampo. Done for today but for future reference: https://github.com/WordPress/gutenberg/actions/runs/10841830322/job/30086485959?pr=65175

**Comment by tyxla** [comment]
Disregarding the failing tests, the problems with the missing indicator looks fixed in my testing 👍

**Comment by ciampo** [comment]
Yup, unit test failures seem related to the changes in this PR.

File: packages/components/src/toggle-group-control/toggle-group-control/styles.ts
Comment (ciampo) [reply packages/components/src/toggle-group-control/toggle-group-control/styles.ts]: > but unfortunately it distorts border radius values. You can play with my animation sandbox here to see what I mean: [stackblitz.com/edit/solidjs-templates-grt7kw?file=src%2FApp.tsx](https://stackblitz.com/edit/solidjs-templates-grt7kw?file=src%2FApp.tsx)

I understand the issue that you're describing, although I wasn't able to trigger it in your demo.

As a follow-up, we could look at https://css-tricks.com/various-methods-for-expanding-a-box-while-preserving-the-border-radius/#aa-changing-scale so that we could use `transform`

**Comment by DaniGuardiola** [comment]
After a very annoying journey into the rabbit hole of "why would this fail in unit tests but not in the real thing", I gave up on making these effects/derived state work properly and I just derive the value as needed where needed since it's only necessary for `indicatorPosition`. It works, so 🤷.

Please re-review and let's get this merged :) @ciampo

File: packages/components/src/toggle-group-control/toggle-group-control/styles.ts
Comment (DaniGuardiola) [reply packages/components/src/toggle-group-control/toggle-group-control/styles.ts]: Yeah, I'd like to look into it, I even had a chat with Matt Perry (Framer Motion author) who pointed me to the place in the codebase where these corrections happen. But yeah, let's better do it in a follow up. After the upcoming refactor, vertical tabs will potentially benefit from that technique too.

**Comment by ciampo** [comment]
The component tests well for me in Storybook, even if the animation's easing is definitely different. The one in trunk seems to have a stronger easing, while the one in this PR feels more "linear".

I then tested the component in the editor. I added a paragraph and tried to change the font size. The animation is clunkier, and very often the backdrop just "skips" to its new position, instead of animating. Not sure if it's because of the CPU being busy with applying typography changes in the canvas, or a "pure" bug. Here's a screen recording:

https://github.com/user-attachments/assets/ec5f72e5-8b15-4ce8-b471-f3c93d236727

**Comment by DaniGuardiola** [comment]
@ciampo oh, good catch. I'd tested other instances but not this one.

The animation is absolutely smooth for me on every instance, including this one. However, it does seem to skip every other change. Very strange. Looking into it, it might be that React reconciliation is being messed with somehow.

**Comment by DaniGuardiola** [comment]
@ciampo update, I managed to fix it by doing things more directly, see my latest commits. It doesn't skip now.

This also feels better in general. Depending on React scheduling is tricky for things like this, so this should be more robust. Another advantage I just realized is that this could be neatly abstracted and reused in Tabs in the near future to avoid duplicated code and maintenance, so that's cool.

One last thing I want to tinker with is using a transform again. This implementation is fine for all intents and purposes, but the limitations of transitioning these non-transform CSS properties can be noticed if you intentionally go crazy on the buttons (or fall asleep on one of them) so that you switch between options very fast. I did a quick test and with `transform` it looks smoother. In real usage, this is a non-issue, but it's still bugging me so I think I'm gonna attempt the border radius thing so we can have transform. The math to achieve this from the article you shared (in some other thread where we discussed this) doesn't seem too complex, so worth a try.

In the meantime, please let me know if the fix works for you, and whether I broke something somewhere else with my fix, hopefully not lol. I tried to do a more or less complete round of testing this time but I'm not scared I'll have missed something :P

**Comment by ciampo** [comment]
I couldn't find any more bugs, but I can definitely still perceive some clunkiness and I agree with you that it would be better to use CSS `transform` 👌  Hopefully the math around keeping rounded corners well-proportioned is not too tricky to get right.

**Comment by DaniGuardiola** [comment]
@ciampo I implemented the transform-based transition with the border radius technique. I did a quick "blind test", and if you compare the performance with the previous framer-motion based implementation, it feels the same if not slightly better to me. To make the comparison fair, the duration ([of the current transition](https://github.com/WordPress/gutenberg/blob/8939f1f84015530b794332ea2cf21af2592c798e/packages/components/src/toggle-group-control/toggle-group-control/styles.ts#L33)) needs to be set to `0.3s` and the timing function to `ease-in-out`, as those are the defaults for framer-motion ([duration](https://github.com/framer/motion/blob/51ac271f9766fe8864ff1f1974de92215e483ad0/packages/framer-motion/src/animation/sequence/create.ts#L38) - also see [the docs](https://www.framer.com/motion/transition/###duration), [easing](https://github.com/framer/motion/blob/51ac271f9766fe8864ff1f1974de92215e483ad0/packages/framer-motion/src/animation/sequence/create.ts#L30)).

I also managed to break it in a new way so there's a fix for that too. Let me know if you can reproduce my results!

**Comment by DaniGuardiola** [comment]
Heads up, I've DRYed it up a bit by moving the animation code into a `useSubelementAnimation` hook for simplicity and because we could reuse this in the Tabs animation to de-duplicate the logic (in a follow-up).

**Comment by ciampo** [comment]
The transition feels snappier, and I think that the border radius math is working well!

Although it looks like the deselectable version of the component has a visual bug when un-selecting an option item:

https://github.com/user-attachments/assets/59c06a3e-d886-44de-8dd7-96032953fe4c

It's also probably better to incorporate changes from #64371 , since it introduces tweaks to `packages/components/src/utils/element-rect.ts`. We can do a final round of review then

File: packages/components/src/toggle-group-control/toggle-group-control/styles.ts
Comment (ciampo) [inline packages/components/src/toggle-group-control/toggle-group-control/styles.ts:38]: Do you think that adding `contain: strict` to the indicator could potentially squeeze some extra rendering performance? It probably won't hurt anyway?

Comment (DaniGuardiola) [reply packages/components/src/toggle-group-control/toggle-group-control/styles.ts:38]: Based on what I've read about the property, I'm not sure that'd do anything for us in the indicator. I could imagine _maybe possibly_ doing something in the control container itself, though I doubt there's any tangible benefit and I'm worried that it could cause some headaches or obscure bugs down the road.

I'd prefer not to use it for now, at least until I or someone else becomes knowledgeable enough on this method to determine its effects with confidence, and only if there's a clear benefit.

**Comment by DaniGuardiola** [comment]
@ciampo found the cause, works as expected now. Also, that PR merged and I synced here. Did one final round of testing, all looks good on my side. Final review?

File: packages/components/src/toggle-group-control/toggle-group-control/as-radio-group.tsx
Comment (ciampo) [inline packages/components/src/toggle-group-control/toggle-group-control/as-radio-group.tsx:81]: This mismatch between `null` and `undefined` as possible `value` types was already there prior to this PR, and it sounds like something we should solve — cc @mirka , could this be part of the work you're intending to do about fixing [ToggleGroupControl's focus bug](https://github.com/WordPress/gutenberg/issues/62981)?

File: packages/components/src/toggle-group-control/toggle-group-control/component.tsx
Comment (ciampo) [inline packages/components/src/toggle-group-control/toggle-group-control/component.tsx:41]: This utility can be quite difficult to understand for someone who is not familiar with your work and the component — we should add some code comments to help give more context

File: packages/components/src/toggle-group-control/toggle-group-control/styles.ts
Comment (ciampo) [inline packages/components/src/toggle-group-control/toggle-group-control/styles.ts:58]: I think it could be worth adding a code comment explaining that we need this sort of math to ensure that the border radius doesn't get distorted while scaling the indicator, and pointing to https://css-tricks.com/various-methods-for-expanding-a-box-while-preserving-the-border-radius/#aa-changing-scale as the inspiration for it.

**Review by ciampo — APPROVED** [review]
I pushed a couple of commits to remove merge conflicts, fix snapshots and make sure that I was testing on top of latest `trunk`.

I couldn't find any bugs or regressions, and the animation definitely feels smoother with the recent change to using `transform`.

I left a couple of comments, but they're small stuff and this PR can be merged once they are addressed

LGTM 🚀

File: packages/components/src/toggle-group-control/toggle-group-control/component.tsx
Comment (DaniGuardiola) [reply packages/components/src/toggle-group-control/toggle-group-control/component.tsx:41]: agreed, hard to understand in its current state, this is gonna get a fast follow that takes it to the utils dir, and then another that makes Tabs use it too, so in that process it can be properly documented. I have those changes already locally.

Do you think it's okay to merge as is now with this context, since your concern will be addressed very soon?

File: packages/components/src/toggle-group-control/toggle-group-control/styles.ts
Comment (DaniGuardiola) [reply packages/components/src/toggle-group-control/toggle-group-control/styles.ts:58]: yes, makes sense. I'll do that before merge here.

File: packages/components/src/toggle-group-control/toggle-group-control/component.tsx
Comment (ciampo) [reply packages/components/src/toggle-group-control/toggle-group-control/component.tsx:41]: It doesn't hurt to add comments now, while you're also adding the ones for the CSS border radius calculation. Good to know that there are plans to move the code to a util

Comment (DaniGuardiola) [reply packages/components/src/toggle-group-control/toggle-group-control/component.tsx:41]: Fair, done.

File: packages/components/src/toggle-group-control/toggle-group-control/styles.ts
Comment (DaniGuardiola) [reply packages/components/src/toggle-group-control/toggle-group-control/styles.ts:58]: done, didn't link the article because it's fairly simple to explain "adjusting the border radius to match the scaling in the x axis"

**Pull request merged**
